include/dali/core/access_order.h

```diff
@@ -0,0 +1,115 @@
+
+#include <cstddef>
+#include "dali/core/api_helper.h"
+#include "dali/core/cuda_error.h"
+#include "dali/core/cuda_stream.h"
+
+namespace dali {
+
+/**
+ * @brief Describes the order in which an object is accessed. It can be stream or host.
+ *
+ * The access order can be either a CUDA stream order or host order. There's also
+ * a special "null" value, useful when passing AccessOrder to functions/objects which may have
+ * internal streams.
+ */
+class DLL_PUBLIC AccessOrder {
+ public:
+  constexpr AccessOrder() = default;
+  constexpr AccessOrder(cudaStream_t stream, int device_id)
+  : stream_(stream), device_id_(device_id) {}
+
+  AccessOrder(int) = delete;  // NOLINT  prevent construction from 0
+  AccessOrder(std::nullptr_t) = delete;  // NOLINT
+
+  constexpr AccessOrder(cudaStream_t stream) : stream_(stream) {  // NOLINT
+    if (is_device())
+      device_id_ = DeviceFromStream(stream);
+  }
+
+  static constexpr AccessOrder host() {
+    return AccessOrder(host_sync_stream());
+  }
+
+  cudaStream_t get() const noexcept { return has_value() ? stream_ : 0; }
+
+  cudaStream_t stream() const noexcept { return is_device() ? stream_ : 0; }
+
+  int device_id() const noexcept { return device_id_; }
+
+  bool is_host() const noexcept { return stream_ == host_sync_stream(); }
+
+  bool is_device() const noexcept {
+    return stream_ != host_sync_stream() && stream_ != null_stream();
+  }
+
+  bool has_value() const noexcept { return stream_ != null_stream(); }
+
+  explicit operator bool() const noexcept { return has_value(); }
+
+  static constexpr cudaStream_t host_sync_stream() noexcept {
+    // TODO(michalz): Unify with dali::mm::host_sync
+    const int i = 4321;
+    // C++ standard makes it impossible to have a constexpr pointer with a fixed
+    // numerical value, so we have to resort to some compiler specific tricks.
+    // Clang and GCC support a __builtin_constant_p hack, other compilers
+    // have relaxed constexpr handling.
+#ifdef __GNUC__
+    return __builtin_constant_p((cudaStream_t)i) ? (cudaStream_t)i : (cudaStream_t)i;
+#else
+    return static_cast<cudaStream_t>(static_cast<void *>(static_cast<char*>(nullptr) + i));
+#endif
+  }
+
+  static constexpr cudaStream_t null_stream() noexcept {
+    const int i = -1;
+    // C++ standard makes it impossible to have a constexpr pointer with a fixed
+    // numerical value, so we have to resort to some compiler specific tricks.
+    // Clang and GCC support a __builtin_constant_p hack, other compilers
+    // have relaxed constexpr handling.
+#ifdef __GNUC__
+    return __builtin_constant_p((cudaStream_t)i) ? (cudaStream_t)i : (cudaStream_t)i;
+#else
+    return static_cast<cudaStream_t>(static_cast<void *>(static_cast<char*>(nullptr) - i));
+#endif
+  }
+
+  /**
+   * @brief Waits in `this` ordering context for the work scheduled in the `other` order.
+   */
+  void join(const AccessOrder &other) const;
+
+  void wait(cudaEvent_t event) const;
```

Comment:
Maybe you both should be called wait or join? What is the difference?

Reply:
Definitely `wait`. `join` is used for threads - that's why I used it for streams, as per analogy, but if you insist, I can rename it to `wait`.

Reply:
My point is that we have `wait` and `join` and I as a user of the API I don't know which one I should use and what is the difference, so maybe having just one with a name with different argument types would be better.